src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp

```diff
@@ -0,0 +1,212 @@
+  // decrease p refinement in that dimension.
+  tmpl::for_each<TensorTags>([&result, &box, &mesh, &power_monitors,
+                              target_abs_truncation_error =
+                                  target_abs_truncation_error_,
+                              target_rel_truncation_error =
+                                  target_rel_truncation_error_](
+                                 const auto tag_v) {
+    // Stop if we have already decided to refine every dimension
+    if (result == make_array<Dim>(Flag::IncreaseResolution)) {
+      return;
+    }
+    using tag = tmpl::type_from<std::decay_t<decltype(tag_v)>>;
+    const auto& tensor = db::get<tag>(box);
+    for (const DataVector& tensor_component : tensor) {
+      PowerMonitors::power_monitors(make_not_null(&power_monitors),
+                                    tensor_component, mesh);
+      for (size_t d = 0; d < Dim; ++d) {
+        // Skip this dimension if we have already decided to refine it
+        if (result[d] == Flag::IncreaseResolution) {
+          continue;
+        }
+        const auto& modes = gsl::at(power_monitors, d);
+        // Increase p refinement if the truncation error exceeds the target
+        const double rel_truncation_error = pow(
+            10, -PowerMonitors::relative_truncation_error(modes, modes.size()));
+        const double umax = max(abs(tensor_component));
+        const double abs_truncation_error = umax * rel_truncation_error;
+        if ((target_rel_truncation_error.has_value() and
+             rel_truncation_error > target_rel_truncation_error.value()) or
+            (target_abs_truncation_error.has_value() and
+             abs_truncation_error > target_abs_truncation_error.value())) {
+          result[d] = Flag::IncreaseResolution;
+          continue;
+        }
+        // Dont' check if we want to (allow) decreasing p refinement if another
+        // tensor has already decided that decreasing p refinement is bad.
+        if (result[d] == Flag::DoNothing) {
+          continue;
+        }
+        // The `PowerMonitors::relative_truncation_error` function requires at
+        // least two modes (and we subtract one below)
+        if (modes.size() >= 3) {
+          // Decrease p refinement if the truncation error with the highest mode
+          // removed still satisfies the target. Otherwise, request to stay at
+          // this resolution (or increase resolution if another tensor component
+          // requested that).
+          const double rel_truncation_error_coarsened =
+              pow(10, -PowerMonitors::relative_truncation_error(
+                          modes, modes.size() - 1));
+          const double abs_truncation_error_coarsened =
+              umax * rel_truncation_error_coarsened;
+          if ((target_rel_truncation_error.has_value() and
+               rel_truncation_error_coarsened <=
+                   target_rel_truncation_error.value()) and
```

Comment:
Tried it but I think it makes the logic a bit harder to understand because the two conditions for refine/coarsen have to use different `value_or` values.